### ISSUE TITLE
Restart not working as expected.

### ISSUE THREAD

**Issue by indexzero** [LearnBoost/cluster#29]
Seeing some unexpected behavior when attempting to restart master / workers programatically. I've summarized the problem here: https://gist.github.com/e4ef8fcbb5be2f468967

Is this an issue or am I doing something wrong?

**Comment by rauchg** [comment]
Can you try with node `v0.4.1` ?

**Comment by indexzero** [comment]
Already using 0.4.1

**Comment by tj** [comment]
A few others had issues too, I just need to reproduce them. One gotcha that I need to add to thendocs is reload() with __dirname can fail due to cluster writing files in that dir, causing a cyclic restart

**Comment by tj** [comment]
I might just have the plugins expose files they work with so reload can skip them

**Comment by tj** [comment]
I should note too that if you `restart()`, and workers have connections they will have to wait until connections have closed, or the timeout. if needed you can force a hard restart by passing a different signal `restart('SIGTERM')`

**Comment by tj** [comment]
another note, for 4 workers your script there is actually executed 5 times, so `master.restart()` from a worker wont work, but you can do:

```
  exports.cluster = cluster(server).listen(80);

  // some file
  if (cluster.isMaster) cluster.restart();
```

let me know if either of those help (isChild, and isWorker are available as well)

**Comment by indexzero** [comment]
I set the timeout to 100ms and I was waiting that long before checking the process list again. The problem is the incoming request could be handled by any of the processes in the cluster, either master or worker. Is there someway to force a restart of all workers from another worker? Maybe through clever usage of the repl plugin?

**Comment by tj** [comment]
master does not listenFD() / accept(), only the workers do, BUT regardless the newly spawned master should handling the connections even if the others are busy. More details of when you are calling `restart()` might help, I will try and reproduce it

**Comment by indexzero** [comment]
Basically I have a route called `/refresh` which when called should restart the server. I'm really just trying to refresh the cache across all processes, but I'm not quite sure the best way to do custom IPC.

**Comment by tj** [comment]
hmm. hope it's a secured route :)

Guess it really depends what you are doing, but you can loop `master.children` and do `worker.call('methodName')`, extending `Worker`, although there is no explicit public api for that right now really

**Comment by tj** [comment]
BUT that being said, if your other process knows the master's PID, you can just `process.kill(masterPID, 'SIGUSR2')` to gracefully restart. I dont have a signal mapped to a hard restart right now

**Comment by tj** [comment]
ah, just realized I totally know what your issue is. if you are calling restart() within a route, you _are_ within a worker, so `Master#restart()` wont work as expected, most of the methods are not callable from within a worker, all of the plugins (so far) act within the master process. but you can still get away with sending a signal to `process.env.CLUSTER_MASTER_PID`

**Comment by tj** [comment]
should be fixed now, found a pretty nasty bug fixed by 5bc31b4
